lib/stages/ringmap.cpp

```diff
@@ -0,0 +1,586 @@
+
+    // Initialize the time indexing
+    min_ctime = ts_to_double(std::get<1>(in_frame.time));
+    max_ctime = min_ctime;
+    latest = modulo<size_t>(num_time);
+
+    return true;
+}
+
+void mapMaker::gen_matrices() {
+
+    // calculate baseline for every stacked product
+    ns_baselines.reserve(num_bl);
+    chimeFeed input_a, input_b;
+    for (size_t i = 0; i < num_bl; i++) {
+        stack_ctype s = stacks[i];
+        input_a = chimeFeed::from_input(inputs[prods[s.prod].input_a]);
+        input_b = chimeFeed::from_input(inputs[prods[s.prod].input_b]);
+        ns_baselines[i] = feed_sep * (input_b.feed_location - input_a.feed_location);
+        if (s.conjugate)
+            ns_baselines[i] *= -1;
+    }
+
+    // Construct matrix of phase weights for every baseline and pixel
+    for (auto f : freqs) {
+        std::vector<cfloat> m(num_pix * num_bl);
+        float lam = wl(f.second.centre);
+        for (uint p = 0; p < num_pix; p++) {
+            for (uint i = 0; i < num_bl; i++) {
+                m[p * num_bl + i] =
+                    std::exp(cfloat(-2.i) * pi * ns_baselines[i] / lam * sinza[p]) / float(num_bl);
+            }
+        }
+        vis2map.insert(std::pair<uint64_t, std::vector<cfloat>>(f.first, m));
+    }
+}
+
+int64_t mapMaker::resolve_time(time_ctype t) {
+
+    if (t.ctime < min_ctime) {
+        // time is too old, discard
+        WARN("Frame older than oldest time in ringmap. Discarding.");
+        return -1;
+    }
+
+    if (t.ctime > max_ctime) {
+        mtx.lock();
+        // We need to add a new time
+        max_ctime = t.ctime;
+        // Increment position
+        if (times_map.size() < num_time) {
+            // Still filling in the array
+            if (times_map.size() > 0)
+                latest++;
+            times.push_back(t);
+        } else {
+            // Remove oldest entry
+            min_ctime = times[++latest].ctime;
+            times_map.erase(min_ctime);
+            times[latest] = t;
+        }
+        times_map.insert(std::pair<double, size_t>(t.ctime, latest));
+        // Clear maps
+        size_t start = latest;
+        size_t stop = size_t(latest) + 1;
+        for (auto f : freqs) {
+            uint64_t fid = f.first;
+            for (uint p = 0; p < num_pol; p++) {
+                std::fill(map.at(fid).at(p).begin() + start * num_pix,
+                          map.at(fid).at(p).begin() + stop * num_pix, 0.);
+                std::fill(wgt.at(fid).at(p).begin() + start, wgt.at(fid).at(p).begin() + stop, 0.);
+            }
+        }
+        mtx.unlock();
+
+        return latest;
+    }
+
+    // Otherwise find the existing time
+    auto res = times_map.find(t.ctime);
+    if (res == times_map.end()) {
+        // No entry for this time
+        WARN("Could not find this time in ringmap. Discarding.");
+        return -1;
+    }
+    return res->second;
+}
+
+redundantStack::redundantStack(Config& config, const string& unique_name,
```

Comment:
It looks like this has a large amount in common with the `baselineCompression` stage, and so I think we can probably merge these two. I don't think you need to do much beyond teach that stage how to work from the stack dataset and not just the prod dataset, and then write a new stack type definition function.

Reply:
That's right. Initially I was hesitating between building this into `baselineCompression` or just emulating it and I went this way to avoid interfering with the ciritical code. But I agree it would make sense to merge them.